pkg/slicegwrecycler/recycler.go

```diff
@@ -0,0 +1,207 @@
+	log := logger.FromContext(ctx).WithName("slicegw")
+	sliceGwName := sliceGw.Name
+	foundsvc := &corev1.Service{}
+	var sliceGwNodePorts []int
+	no, _ := getNumberOfGatewayNodePortServices(ctx, meshClient, sliceGw)
+	if no != n {
+		// capping the number of services to be 2 for now, later i will be equal to number of gateway nodes,eg: i = len(node.GetExternalIpList())
+		for i := 0; i < n; i++ {
+			err := meshClient.Get(ctx, types.NamespacedName{Name: "svc-" + sliceGwName + "-" + fmt.Sprint(i), Namespace: controllers.ControlPlaneNamespace}, foundsvc)
+			if err != nil {
+				if errors.IsNotFound(err) {
+					svc := serviceForGateway(sliceGw, i)
+					log.Info("Creating a new Service", "Namespace", svc.Namespace, "Name", svc.Name)
+					err = meshClient.Create(ctx, svc)
+					if err != nil {
+						log.Error(err, "Failed to create new Service", "Namespace", svc.Namespace, "Name", svc.Name)
+						return true, ctrl.Result{}, nil, err
+					}
+					return true, ctrl.Result{Requeue: true}, nil, nil
+				}
+				log.Error(err, "Failed to get Service")
+				return true, ctrl.Result{}, nil, err
+			}
+		}
+	}
+	sliceGwNodePorts, _ = getNodePorts(ctx, meshClient, sliceGw)
+	err := hubClient.UpdateNodePortForSliceGwServer(ctx, sliceGwNodePorts, sliceGwName)
+	if err != nil {
+		log.Error(err, "Failed to update NodePort for sliceGw in the hub")
+		utils.RecordEvent(ctx, eventRecorder, sliceGw, nil, ossEvents.EventSliceGWNodePortUpdateFailed, controllerName)
+		return true, ctrl.Result{}, sliceGwNodePorts, err
+	}
+	return false, reconcile.Result{}, sliceGwNodePorts, nil
+}
+
+func createHeadlessServiceForGwServer(slicegateway *kubeslicev1beta1.SliceGateway) *corev1.Service {
+	svc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      slicegateway.Status.Config.SliceGatewayRemoteGatewayID,
+			Namespace: slicegateway.Namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports:     nil,
+			Selector:  nil,
+			ClusterIP: "None",
+		},
+	}
+	// ctrl.SetControllerReference(slicegateway, svc, r.Scheme)
+	return svc
+}
+
+// getRemoteGwPodName returns the remote gw PodName.
+func getRemoteGwPodName(ctx context.Context, meshClient client.Client, gwRemoteVpnIP string, podIP string) (string, error) {
+	log := logger.FromContext(ctx)
+	log.Info("calling gw sidecar to get PodName", "type", "slicegw")
+	sidecarGrpcAddress := podIP + ":5000"
+	workerGWClient, err := sidecar.NewWorkerGWSidecarClientProvider()
+	if err != nil {
+		os.Exit(1)
+	}
+	remoteGwPodName, err := workerGWClient.GetSliceGwRemotePodName(ctx, gwRemoteVpnIP, sidecarGrpcAddress)
+	log.Info("slicegw remote pod name", "slicegw", remoteGwPodName)
+	if err != nil {
+		log.Error(err, "Failed to get slicegw remote pod name. PodIp: %v", podIP)
+		return "", err
+	}
+	return remoteGwPodName, nil
+}
+
+func getNumberOfGatewayNodePortServices(ctx context.Context, meshClient client.Client, sliceGw *kubeslicev1beta1.SliceGateway) (int, error) {
```

Comment:
Can all this function be an receiver of struct recyclerClient so that we avoid passing all these repeat arguments